lib/jsonWebToken.js

```diff
@@ -0,0 +1,186 @@
+   * Checking jwtString structure, getting header, payload and signature
+   ********************************************************************/
+
+  // split jwtString, make sure we have three parts and we have signature
+  var parts = jwtString.split('.');
+  if (parts.length !== 3)
+    return done(new Error('jwtString is malformed'));
+  if (parts[2] === '')
+    return done(new Error('signature is missing in jwtString'));
+  
+  // decode jwsString
+  var decodedToken;
+  try {
+    decodedToken = jws.decode(jwtString);
+  } catch(err) {
+    return done(new Error('invalid token'));
+  }
+
+  if (!decodedToken) {
+    return done(new Error('invalid token'));
+  }
+
+  // get header, payload and signature
+  var header = decodedToken.header;
+  var payload = decodedToken.payload;
+  var signature = decodedToken.signature;
+
+  if (!header)
+    return done(new Error('missing header in the token'));
+  if (!payload)
+    return done(new Error('missing payload in the token'));
+  if (!signature)
+    return done(new Error('missing signature in the token'));
+
+  /*********************************************************************
+   * validate algorithm and signature
+   ********************************************************************/
+
+  // alg in the token header should be one of those provided in options.algorithms
+  if (!~options.algorithms.indexOf(header.alg)) {
+    return done(new Error('invalid algorithm'));
+  }
+
+  try {
+    var valid = jws.verify(jwtString, header.alg, PMEKey);
+    if (!valid)
+      return done(new Error('invalid signature'));
+  } catch (e) {
+    return done(e);
+  }
+
+  /*********************************************************************
+   * validate payload content
+   ********************************************************************/
+
+  // (1) issuer
+  //   - check the existence and the format of payload.iss
+  //   - validate if options.issuer is set
+  if (typeof payload.iss !== 'string')
+    return done(new Error('invalid iss value in payload'));
+  if (options.issuer && options.issuer !== payload.iss)
+    return done(new Error('jwt issuer is invalid. expected: ' + options.issuer));
+
+  // (2) subject
+  //   - check the existence and the format of payload.sub
+  //   - validate if options.subject is set
+  if (typeof payload.sub !== 'string')
+    return done(new Error('invalid sub value in payload'));
+  if (options.subject && options.subject !== payload.sub)
+      return done(new Error('jwt subject is invalid. expected: ' + options.subject));
+
+  // (3) audience
+  //   - always validate
+  //   - allow payload.aud to be an array of audience
+  //   - options.audience must be a string
+  if (typeof options.audience !== 'string')
+    return done(new Error('invalid options.audience value'));
+  var payload_audience = Array.isArray(payload.aud) ? payload.aud : [payload.aud];
+  if (payload_audience.indexOf(options.audience) == -1)
```

Comment:
added check for options.audience, since later we compare with payload.aud, this will suffice.